env.sh

```diff
@@ -1,87 +1,85 @@
+
+# Docker Port
+EXPOSE="1337:1337"
+if [[ "$3" == "port" ]]; then
+    EXPOSE="${*:4}"
+    [[ "$EXPOSE" = "" ]] && EXPOSE="1337:1337"
+    docker container stop $CONTAINER_NAME > /dev/null 2>&1
+    echo -e "\033[0;32mExpose ports: ${*}\033[0m"
 fi
 
+# Docker Arguments
 WORKDIR=/opt/cita
-USER_ID=`id -u $USER`
+USER_ID="$(id -u "$USER")"
 USER_NAME="user"
 
-[[ "${USER_ID}" = "0" ]] && USER_NAME="root"
-
-# Init contanier's cargo, for logs.
-CARGO_HOME=/opt/.cargo
-DOCKER_CARGO=${HOME}/.docker_cargo
-mkdir -p ${DOCKER_CARGO}/git
-mkdir -p ${DOCKER_CARGO}/registry
+cp /etc/localtime "$SOURCE_DIR"/localtime
+LOCALTIME_PATH="$SOURCE_DIR"/localtime
+[[ ${USER_ID} = "0" ]] && USER_NAME="root"
 
 INIT_CMD="sleep infinity"
 
-# Port condition
-EXPOSE="1337:1337"
-if [ "$3" == "port" ]; then
-    EXPOSE=${@:4}
-    [[ "${EXPOSE}" = "" ]] && EXPOSE="1337:1337"
-    docker container stop $CONTAINER_NAME > /dev/null 2>&1
-    echo -e "\033[0;32mExpose ports: ${@}\033[0m"
-fi
-
-# Run container
+# Run Docker
 docker ps | grep ${CONTAINER_NAME} > /dev/null 2>&1
-if [ $? -ne 0 ]; then
-    echo "Start docker container ${CONTAINER_NAME} ..."
+ret="$?"
+if [[ "$ret" -ne 0 ]]; then
+    echo "Start docker container $CONTAINER_NAME ..."
     docker rm ${CONTAINER_NAME} > /dev/null 2>&1
     docker run -d \
            --net=${SYSTEM_NET} \
-           --volume ${SOURCE_DIR}:${WORKDIR} \
-           --volume ${DOCKER_CARGO}/git:${CARGO_HOME}/git \
-           --volume ${DOCKER_CARGO}/registry:${CARGO_HOME}/registry \
-           --volume ${LOCALTIME_PATH}:/etc/localtime \
-           --env USER_ID=${USER_ID} \
+           --volume "$SOURCE_DIR":"$WORKDIR" \
+           --volume "$DOCKER_CARGO"/git:"$CARGO_HOME"/git \
+           --volume "$DOCKER_CARGO"/registry:"$CARGO_HOME"/registry \
+           --volume "$LOCALTIME_PATH":/etc/localtime \
+           --env USER_ID="$USER_ID" \
            --workdir ${WORKDIR} \
            --name ${CONTAINER_NAME} \
-           -p $EXPOSE ${DOCKER_IMAGE} \
-           /bin/bash -c "${INIT_CMD}"
+           -p ${EXPOSE} ${DOCKER_IMAGE} \
+           /bin/bash -c "$INIT_CMD"
     # Wait entrypoint.sh to finish
     sleep 3
 fi
 
-test -t 1 && USE_TTY="-t"
-
-# Start nodes outside container directly
-# Runing Commands through `cita` command:
+# Run commands through docker container
 # $0=`realpath`, $1="bin/cita", $2="bebop",
 # $3="command/--daemon", $4="config/command"
-# Most OS delete $0 default, some linux not.
+test -t 1 && USE_TTY="-t"
 [[ "$3" == "start" ]] || [[ "$3" == "restart" ]] && set "${@:1:2}" "--daemon" "${@:3}"
 
-# Condition `daemon` to run daemon.
-if [ "$3" == "--daemon"  ]; then
+if [[ "$3" == "--daemon" ]]; then
     set "${@:1:2}" "${@:4}"
-    docker exec -d ${CONTAINER_NAME} /bin/bash -c "/usr/bin/gosu ${USER_NAME} $* >/dev/null 2>&1"
-elif [ $# -gt 0 ]; then
+    docker exec -d ${CONTAINER_NAME} /bin/bash -c /usr/bin/gosu ${USER_NAME} "$@" >/dev/null 2>&1
+elif [[ $# -gt 0 ]]; then
```

Comment:
I am not sure about the correct usage `$@` or `$*`. But，we had changed `$@` to `$*` in #401 on purpose.

@yangby-cryptape  Sorry to take you some times to review this command again.

Reply:
I can't start the chain...

Reply:
`ShellCheck` recommends to use `$*` too, as it [says][1]: 

> Assigning an array to a string! Assign as array, or use * instead of @ to concatenate.

+ `$*` for `string`
+ `$@` for separated arguments

I guess.

I used to use `$@` and never using `$*` so I didn't change it in case...

This modification was written last week and I forget to pr XD, so I didn't notice the new change.

The explanation in #401 does make sense, I'll change `$@` to `$*` in docker commands and do more test about it.

[1]: https://github.com/koalaman/shellcheck/wiki/SC2124

Reply:
> Maybe it's better to run the travis that uses `env.sh`.
> And the circleCI uses `bebop`

I like modifying shell commands, and I have to try to test them comprehensively(both installing and running), because the shell scripts are the entrance to cita users, every little change might create big problems to users using cita.

I find I depended on travis overly after I noticed them in my first pr, even though I don't know exactly if travis' scripts covered every tests about my changing.

I always modifying too many lines in shell scripts though I feel stressed with every pr I used to pr...but anyway, I should make `env.sh` and `scripts/cita` performing best in where I had modified.

Reply:
> every little change might create big problems to users using cita.

That's our fault. we should add test of cita usage into CI

> I always modifying too many lines in shell scripts though I feel stressed with every pr I made

Small pr, happy life :sunglasses:

Reply:
Just ignore my ranting 🤦🏻‍♂️

> Small pr, happy life 😎

Might be the truth...

> I can't start the chain...

Just catch the bug,  as #401's patch: 

```shell
# ATTENTION: `-d` required passing full string.                                                                                                                                                           
docker exec -d ${CONTAINER_NAME} /bin/bash -c "/usr/bin/gosu ${USER_NAME} $* >/dev/null 2>&1"
```

+ `docker exec -d` should passing full string: `$*`, because it crosses two tty? and docker does special processes to the `string`? 
+ but `docker exec -i` should use parameters: `$@`, because of the normal passing?

Really a big discovery for me, I used to test `docker exec -d` over and over again in my early pr, and never noticed this question.